src/BoltOn.Data.CosmosDb/BaseCosmosDbRepository.cs

```diff
@@ -0,0 +1,118 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Documents.Client;
+using Microsoft.Azure.Documents.Linq;
+
+namespace BoltOn.Data.CosmosDb
+{
+    public abstract class BaseCosmosDbRepository<TEntity, TCosmosDbContext> : IRepository<TEntity>
+        where TEntity : class
+        where TCosmosDbContext : BaseCosmosDbContext<TCosmosDbContext>
+    {
+        protected readonly string _databaseName;
+        protected readonly string _collectionName;
+        protected readonly DocumentClient _client;
+
+        public BaseCosmosDbRepository(TCosmosDbContext cosmosDbContext, string collectionName = null)
+        {
+            _databaseName = cosmosDbContext.DatabaseName;
+            _collectionName = collectionName ?? typeof(TEntity).Name.Pluralize();
+            _client = cosmosDbContext.DocumentClient;
+        }
+
+        public virtual TEntity Add(TEntity entity)
+        {
+            _client.CreateDocumentAsync(GetDocumentCollectionUri(), entity).GetAwaiter().GetResult();
+            return entity;
+        }
+
+        public virtual async Task<TEntity> AddAsync(TEntity entity, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            await _client.CreateDocumentAsync(GetDocumentCollectionUri(), entity);
+            return entity;
+        }
+
+        public virtual IEnumerable<TEntity> FindBy(Expression<Func<TEntity, bool>> predicate, params Expression<Func<TEntity, object>>[] includes)
+        {
+            var query = _client.CreateDocumentQuery<TEntity>(GetDocumentCollectionUri())
+                .Where(predicate)
+                .AsDocumentQuery();
+
+            return GetResultsFromDocumentQuery(query).GetAwaiter().GetResult();
```

Comment:
@gokulm : read somewhere on stackoverflower microsoft also uses this GetAwaiter().GetResult() to call async methods from sync ones.

Reply:
could lead to deadlocks depending on the consumer.

Reply:
@gokulm : any other better way to make it synchronous?

Reply:
try this var result = AsyncContext.Run(MyAsyncMethod); https://stackoverflow.com/questions/9343594/how-to-call-asynchronous-method-from-synchronous-method-in-c